File: ios/SkyCast/App/AppBackgroundViewModel.swift
import Foundation
import Observation

/// The weather the whole app is painted with.
///
/// The favourite location supplies the background behind every screen that has no weather of its
/// own: METAR, Locations and Settings. Home paints its own background per page and draws over this
/// one, and the Moon tab has a night sky instead.
///
/// The Kotlin twin is `ui/AppBackgroundViewModel.kt`.
@MainActor
@Observable
final class AppBackgroundViewModel {
    private(set) var condition: WeatherCondition = .unknown
    private(set) var isDaytime = true

    private let locationRepository: any LocationRepository
    private let weatherRepository: any WeatherRepository

    private var observationTask: Task<Void, Never>?

    init(locationRepository: any LocationRepository, weatherRepository: any WeatherRepository) {
        self.locationRepository = locationRepository
        self.weatherRepository = weatherRepository
    }

    func start() {
        guard observationTask == nil else { return }
        observationTask = Task { await observe() }
    }

    /// Re-reads the favourite, which may have changed on the Locations tab.
    func reload() async {
        observationTask?.cancel()
        observationTask = Task { await observe() }
    }

    private func observe() async {
        guard let favourite = try? await locationRepository.primaryLocation() else { return }

        // Cached weather satisfies this without a request, Home is already fetching it, and a failure
        // needs no handling. The default is a neutral wash, not an error.
        for await dataState in weatherRepository.currentWeather(for: favourite) {
            if Task.isCancelled {
                return
            }
            guard let weather = dataState.data else { continue }
            condition = weather.condition
            isDaytime = weather.isDaytime
        }
    }
}

File: android/app/src/main/java/com/nauhaan/skycast/ui/AppBackgroundViewModel.kt
package com.nauhaan.skycast.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nauhaan.skycast.domain.model.WeatherCondition
import com.nauhaan.skycast.domain.repository.LocationRepository
import com.nauhaan.skycast.domain.repository.WeatherRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import javax.inject.Inject

/**
 * The weather the whole app is painted with.
 *
 * The favourite location supplies the background behind every screen that has no weather of its
 * own: METAR, Locations and Settings. Home paints its own background per page and draws over this
 * one, and the Moon tab has a night sky instead.
 *
 * A view model rather than a CompositionLocal, because the background is drawn by the shell above
 * the navigation graph, so no screen is in a position to provide it, and it is read once for the
 * whole app.
 */
@HiltViewModel
class AppBackgroundViewModel
@Inject
constructor(
    locationRepository: LocationRepository,
    weatherRepository: WeatherRepository,
) : ViewModel() {
    @OptIn(ExperimentalCoroutinesApi::class)
    val background: StateFlow<AppBackground> = locationRepository
        .observePrimaryLocation()
        .flatMapLatest { favourite ->
            if (favourite == null) {
                flowOf(AppBackground())
            } else {
                // Cached weather satisfies this without a request, since Home is already fetching it, and a
                // failure needs no handling here. `AppBackground()` is a neutral wash, not an error.
                weatherRepository.observeCurrentWeather(favourite).map { state ->
                    val weather = state.data
                    AppBackground(
                        condition = weather?.condition ?: WeatherCondition.UNKNOWN,
                        isDaytime = weather?.isDaytime ?: true,
                    )
                }
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(SUBSCRIPTION_TIMEOUT_MILLIS),
            initialValue = AppBackground(),
        )

    private companion object {
        const val SUBSCRIPTION_TIMEOUT_MILLIS = 5_000L
    }
}

/** The favourite's condition, or a neutral default before one is known. */
data class AppBackground(val condition: WeatherCondition = WeatherCondition.UNKNOWN, val isDaytime: Boolean = true)
